proxmox/resource_cloud_init.go

```diff
@@ -0,0 +1,174 @@
+				Optional: true,
+				ForceNew: true,
+			},
+			"meta_data": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"network_config": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"sha256": {
+				Type:     schema.TypeString,
+				Computed: true,
+				ForceNew: true,
+			},
+			"size": {
+				Type:     schema.TypeString,
+				Computed: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func createCloudInitISO(metaData, userData, networkConfig string) (io.Reader, string, error) {
+	isoWriter, err := iso9660.NewWriter()
+	if err != nil {
+		return nil, "", err
+	}
+	defer isoWriter.Cleanup()
+
+	err = isoWriter.AddFile(strings.NewReader(metaData), "meta-data")
+	if err != nil {
+		return nil, "", err
+	}
+
+	err = isoWriter.AddFile(strings.NewReader(userData), "user-data")
+	if err != nil {
+		return nil, "", err
+	}
+
+	err = isoWriter.AddFile(strings.NewReader(networkConfig), "network-config")
+	if err != nil {
+		return nil, "", err
+	}
+
+	var b bytes.Buffer
+	err = isoWriter.WriteTo(&b, "cidata")
+	if err != nil {
+		return nil, "", err
+	}
+
+	// Calculate the ISO sha256 sum
+	sum := fmt.Sprintf("%x", sha256.Sum256(b.Bytes()))
+
+	return bytes.NewReader(b.Bytes()), sum, nil
+}
+
+func resourceCloudInitDiskCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	pconf := m.(*providerConfiguration)
+	client := pconf.Client
+
+	r, sum, err := createCloudInitISO(d.Get("meta_data").(string), d.Get("user_data").(string), d.Get("network_config").(string))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	fileName := fmt.Sprintf("tf-ci-%s.iso", d.Get("name").(string))
+	err = client.Upload(d.Get("pve_node").(string), d.Get("storage").(string), isoContentType, fileName, r)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.Set("sha256", sum)
+
+	// The volume ID is the storage name and the file name
+	volId := fmt.Sprintf("%s:%s/%s", d.Get("storage").(string), isoContentType, fileName)
+	d.SetId(volId)
+
+	return resourceCloudInitDiskRead(ctx, d, m)
+}
+
+func resourceCloudInitDiskRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	pconf := m.(*providerConfiguration)
+	client := pconf.Client
+
+	var isoFound bool
+	pveNode := d.Get("pve_node").(string)
+	vmRef := &proxmox.VmRef{}
+	vmRef.SetNode(pveNode)
+	vmRef.SetVmType("qemu")
+	storageContent, err := client.GetStorageContent(vmRef, d.Get("storage").(string))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	contents := storageContent["data"].([]interface{})
+	for c := range contents {
+		storageContentMap := contents[c].(map[string]interface{})
+		if storageContentMap["volid"].(string) == d.Id() {
+			size := storageContentMap["size"].(float64)
+			d.Set("size", ByteCountIEC(int64(size)))
+			isoFound = true
+			break
+		}
+	}
+
+	if !isoFound {
+		// ISO not found so we (re)create it
+		d.SetId("")
+		return nil
+	}
+
+	return nil
```

Comment:
I would return diag and not nil when possible

Reply:
as it's not a an error. We should just return nil right?

Reply:
that should be ok, you're right